LoopKit/GlucoseKit/CachedGlucoseObject+CoreDataClass.swift

```diff
@@ -21,7 +21,20 @@ class CachedGlucoseObject: NSManagedObject {
         set {
             willChangeValue(forKey: "syncVersion")
             defer { didChangeValue(forKey: "syncVersion") }
-            primitiveSyncVersion = newValue != nil ? NSNumber(value: newValue!) : nil
+            primitiveSyncVersion = newValue.map { NSNumber(value: $0) }
+        }
+    }
+    
+    var device: HKDevice? {
+        get {
+            willAccessValue(forKey: "device")
+            defer { didAccessValue(forKey: "device") }
+            return primitiveDevice.flatMap { try? NSKeyedUnarchiver.unarchivedObject(ofClass: HKDevice.self, from: $0) }
```

Comment:
Serializing the device into a relational column isn't great. We did it in DosingDecisionStore, as there Loop specific things that don't belong in LoopKit, and we don't imagine querying dosing decisions in the near term. I think you're right that this would be a good candidate for denormalization; especially because we'll likely just keep storing the same device over and over. However, there isn't a strict need for us to query this column yet, so I won't block this because of that.

Reply:
I was a bit concerned about this as well.  And I came up with the same conclusion: this is just stuffed in CoreData only for rehydration for HealthKit (well, also for uploading to Services if we want).  But we kinda have no choice, we do have to store it (IMHO it is a good thing that the device details are available to the user in Apple Health). Point is: we don't query or index on it, so I thought a serialized format was fine.  I measured, for example, the size of Dexcom's `HKDevice`, and it winds up being ~400 bytes.  Not gigantic, but not tiny either.  If we want to optimize this, we should first figure out what we want to optimize (database size? processing speed? upload bandwidth? something else?).  If we decide size needs to be optimized, I think denormalization would be pretty straightforward.